File: cocos/core/utils/path.js
/****************************************************************************
 Copyright (c) 2013-2016 Chukong Technologies Inc.
 Copyright (c) 2017-2018 Xiamen Yaji Software Co., Ltd.

 http://www.cocos.com

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated engine source code (the "Software"), a limited,
  worldwide, royalty-free, non-assignable, revocable and non-exclusive license
 to use Cocos Creator solely to develop games on your target platforms. You shall
  not use Cocos Creator software for developing other software or tools that's
  used for developing games. You are not granted to publish, distribute,
  sublicense, and/or sell copies of Cocos Creator.

 The software or tools in this License Agreement are licensed, not sold.
 Xiamen Yaji Software Co., Ltd. reserves all rights not expressly granted to you.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
 ****************************************************************************/

const EXTNAME_RE = /(\.[^\.\/\?\\]*)(\?.*)?$/;
const DIRNAME_RE = /((.*)(\/|\\|\\\\))?(.*?\..*$)?/;
const NORMALIZE_RE = /[^\.\/]+\/\.\.\//;

/**
 * !#en The module provides utilities for working with file and directory paths
 * !#zh 用于处理文件与目录的路径的模块
 * @class path
 * @static
 */

/**
 * !#en Join strings to be a path.
 * !#zh 拼接字符串为 Path
 * @method join
 * @example {@link cocos2d/core/utils/CCPath/join.js}
 * @returns {String}
 */
export function join () {
    let l = arguments.length;
    let result = "";
    for (let i = 0; i < l; i++) {
        result = (result + (result === "" ? "" : "/") + arguments[i]).replace(/(\/|\\\\)$/, "");
    }
    return result;
}

/**
 * !#en Get the ext name of a path including '.', like '.png'.
 * !#zh 返回 Path 的扩展名，包括 '.'，例如 '.png'。
 * @method extname
 * @example {@link cocos2d/core/utils/CCPath/extname.js}
 * @param {String} pathStr
 * @returns {*}
 */
export function extname (pathStr) {
    let temp = EXTNAME_RE.exec(pathStr);
    return temp ? temp[1] : '';
}

/**
 * !#en Get the main name of a file name
 * !#zh 获取文件名的主名称
 * @method mainFileName
 * @param {String} fileName
 * @returns {String}
 * @deprecated
 */
export function mainFileName (fileName) {
    if (fileName) {
        let idx = fileName.lastIndexOf(".");
        if (idx !== -1)
            return fileName.substring(0, idx);
    }
    return fileName;
}

/**
 * !#en Get the file name of a file path.
 * !#zh 获取文件路径的文件名。
 * @method basename
 * @example {@link cocos2d/core/utils/CCPath/basename.js}
 * @param {String} pathStr
 * @param {String} [extname]
 * @returns {*}
 */
export function basename (pathStr, extname) {
    let index = pathStr.indexOf("?");
    if (index > 0) pathStr = pathStr.substring(0, index);
    let reg = /(\/|\\)([^\/\\]+)$/g;
    let result = reg.exec(pathStr.replace(/(\/|\\)$/, ""));
    if (!result) return null;
    let baseName = result[2];
    if (extname && pathStr.substring(pathStr.length - extname.length).toLowerCase() === extname.toLowerCase())
        return baseName.substring(0, baseName.length - extname.length);
    return baseName;
}

/**
 * !#en Get dirname of a file path.
 * !#zh 获取文件路径的目录名。
 * @method dirname
 * @example {@link cocos2d/core/utils/CCPath/dirname.js}
 * @param {String} pathStr
 * @returns {*}
 */
export function dirname (pathStr) {
    let temp = DIRNAME_RE.exec(pathStr);
    return temp ? temp[2] : '';
}

/**
 * !#en Change extname of a file path.
 * !#zh 更改文件路径的扩展名。
 * @method changeExtname
 * @example {@link cocos2d/core/utils/CCPath/changeExtname.js}
 * @param {String} pathStr
 * @param {String} [extname]
 * @returns {String}
 */
export function changeExtname (pathStr, extname) {
    extname = extname || "";
    let index = pathStr.indexOf("?");
    let tempStr = "";
    if (index > 0) {
        tempStr = pathStr.substring(index);
        pathStr = pathStr.substring(0, index);
    }
    index = pathStr.lastIndexOf(".");
    if (index < 0) return pathStr + extname + tempStr;
    return pathStr.substring(0, index) + extname + tempStr;
}

/**
 * !#en Change file name of a file path.
 * !#zh 更改文件路径的文件名。
 * @example {@link cocos2d/core/utils/CCPath/changeBasename.js}
 * @param {String} pathStr
 * @param {String} basename
 * @param {Boolean} [isSameExt]
 * @returns {String}
 */
export function changeBasename (pathStr, basename, isSameExt) {
    if (basename.indexOf(".") === 0) return this.changeExtname(pathStr, basename);
    let index = pathStr.indexOf("?");
    let tempStr = "";
    let ext = isSameExt ? this.extname(pathStr) : "";
    if (index > 0) {
        tempStr = pathStr.substring(index);
        pathStr = pathStr.substring(0, index);
    }
    index = pathStr.lastIndexOf("/");
    index = index <= 0 ? 0 : index + 1;
    return pathStr.substring(0, index) + basename + ext + tempStr;
}

//todo make public after verification
export function _normalize (url) {
    let oldUrl = url = String(url);

    //removing all ../
    do {
        oldUrl = url;
        url = url.replace(NORMALIZE_RE, "");
    } while (oldUrl.length !== url.length);
    return url;
}

// @param {string} path
export function stripSep (path) {
    return path.replace(/[\/\\]$/, '');
}

cc.path = {
    join,
    extname,
    mainFileName,
    basename,
    dirname,
    changeExtname,
    changeBasename,
    _normalize,
    stripSep,
    get sep () {
        return (cc.sys.os === cc.sys.OS_WINDOWS ? '\\' : '/');
    }
};

File: native/cocos/core/utils/Path.cpp
/****************************************************************************
 Copyright (c) 2021-2023 Xiamen Yaji Software Co., Ltd.

 http://www.cocos.com

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights to
 use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 of the Software, and to permit persons to whom the Software is furnished to do so,
 subject to the following conditions:

 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
****************************************************************************/

#include "core/utils/Path.h"
#include <cstring>
#include "base/StringUtil.h"

namespace cc {

namespace {
const ccstd::string EMPTY_STRING;

ccstd::string &removeLastSlash(ccstd::string &path) {
    if (!path.empty()) {
        if (path[path.length() - 1] == '/' || path[path.length() - 1] == '\\') {
            path = path.substr(0, path.length() - 1);
        } else if (path.length() > 1 && path[path.length() - 1] == '\\' && path[path.length() - 2] == '\\') {
            path = path.substr(0, path.length() - 2);
        }
    }
    return path;
}

} // namespace

ccstd::string join(const ccstd::vector<ccstd::string> &segments) {
    ccstd::string result;

    for (const auto &segment : segments) {
        if (!result.empty()) {
            result += "/";
        }
        result += segment;
        removeLastSlash(result);
    }

    return result;
}

ccstd::string extname(const ccstd::string &path) {
    if (path.empty()) {
        return EMPTY_STRING;
    }

    ccstd::string newPath = path;
    size_t index = path.find_first_of('?');
    if (index != ccstd::string::npos && index > 0) {
        newPath = newPath.substr(0, index);
    }

    index = newPath.find_last_of('.');
    if (index == ccstd::string::npos) {
        return EMPTY_STRING;
    }

    return newPath.substr(index);
}

ccstd::string mainFileName(const ccstd::string &fileName) {
    if (!fileName.empty()) {
        size_t idx = fileName.find_last_of('.');
        if (idx != ccstd::string::npos) {
            return fileName.substr(0, idx);
        }
    }

    return fileName;
}

ccstd::string basename(const ccstd::string &path, const ccstd::string &extName /* = ""*/) {
    ccstd::string newPath = path;
    size_t index = path.find_first_of('?');
    if (index != ccstd::string::npos && index > 0) {
        newPath = newPath.substr(0, index);
    }

    removeLastSlash(newPath);

    index = newPath.find_last_of("/\\");
    if (index == ccstd::string::npos) {
        return newPath;
    }

    ccstd::string baseName = newPath.substr(index + 1);

    if (!extName.empty() && extName.length() < newPath.length()) {
        ccstd::string extInPath = newPath.substr(newPath.length() - extName.length());
        ccstd::string expectedExtName = extName;
        if (StringUtil::tolower(extInPath) == StringUtil::tolower(expectedExtName)) {
            baseName = baseName.substr(0, baseName.length() - extName.length());
        }
    }

    return baseName;
}

ccstd::string dirname(const ccstd::string &path) {
    size_t index = path.find_last_of("/\\");
    if (index == ccstd::string::npos) {
        return "";
    }

    ccstd::string dir = path.substr(0, index);
    removeLastSlash(dir);
    return dir;
}

ccstd::string changeExtname(const ccstd::string &path, const ccstd::string &extName /* = ""*/) {
    size_t index = path.find_first_of('?');
    ccstd::string newPath = path;
    ccstd::string tempStr;
    if (index != ccstd::string::npos && index > 0) {
        tempStr = path.substr(index);
        newPath = path.substr(0, index);
    }

    index = newPath.find_last_of('.');
    if (index == ccstd::string::npos) {
        return newPath + extName + tempStr;
    }

    return newPath.substr(0, index) + extName + tempStr;
}

ccstd::string changeBasename(const ccstd::string &path, const ccstd::string &baseName, bool isSameExt /* = false*/) {
    if (baseName.find_last_of('.') == 0) {
        return changeExtname(path, baseName);
    }

    size_t index = path.find_last_of('?');
    ccstd::string tempStr;
    ccstd::string newPath = path;
    const ccstd::string ext = isSameExt ? extname(path) : "";
    if (index != ccstd::string::npos && index > 0) {
        tempStr = path.substr(index);
        newPath = path.substr(0, index);
    }

    index = newPath.find_last_of("/\\");
    if (index == ccstd::string::npos) {
        index = 0;
    } else if (index > 0) {
        ++index;
    }

    return newPath.substr(0, index) + baseName + ext + tempStr;
}

ccstd::string normalize(const ccstd::string &url) {
    ccstd::string oldUrl = url;
    ccstd::string newUrl = url;

    // remove all ../
    do {
        oldUrl = newUrl;
        size_t index = newUrl.find("../");
        if (index == ccstd::string::npos) {
            index = newUrl.find("..\\");
        }
        size_t previousSlashIndex = ccstd::string::npos;
        size_t previousTwiceSlashIndex = ccstd::string::npos;
        if (index != ccstd::string::npos && index > 0) {
            previousSlashIndex = newUrl.find_last_of("/\\", index - 1);
            if (previousSlashIndex != ccstd::string::npos) {
                previousTwiceSlashIndex = newUrl.find_last_of("/\\", previousSlashIndex - 1);
            }
        }

        if (previousTwiceSlashIndex == ccstd::string::npos) {
            if (previousSlashIndex != ccstd::string::npos) {
                newUrl = newUrl.substr(index + strlen("../"));
            }
        } else if (previousSlashIndex != ccstd::string::npos) {
            newUrl = newUrl.substr(0, previousTwiceSlashIndex) + '/' + newUrl.substr(index + strlen("../"));
        }
    } while (oldUrl.length() != newUrl.length());

    return newUrl;
}

ccstd::string stripSep(const ccstd::string &path) {
    ccstd::string result = path;
    removeLastSlash(result);
    return result;
}

char getSeperator() {
#if (CC_PLATFORM == CC_PLATFORM_WINDOWS)
    return '\\';
#else
    return '/';
#endif
}

} // namespace cc
